src/core/lib/security/credentials/tls/spiffe_credentials.cc

```diff
@@ -0,0 +1,123 @@
+
+grpc_core::RefCountedPtr<grpc_channel_security_connector>
+grpc_tls_spiffe_credentials::create_security_connector(
+    grpc_core::RefCountedPtr<grpc_call_credentials> call_creds,
+    const char* target_name, const grpc_channel_args* args,
+    grpc_channel_args** new_args) {
+  const char* overridden_target_name = nullptr;
+  tsi_ssl_session_cache* ssl_session_cache = nullptr;
+  for (size_t i = 0; args && i < args->num_args; i++) {
+    grpc_arg* arg = &args->args[i];
+    if (strcmp(arg->key, GRPC_SSL_TARGET_NAME_OVERRIDE_ARG) == 0 &&
+        arg->type == GRPC_ARG_STRING) {
+      overridden_target_name = arg->value.string;
+    }
+    if (strcmp(arg->key, GRPC_SSL_SESSION_CACHE_ARG) == 0 &&
+        arg->type == GRPC_ARG_POINTER) {
+      ssl_session_cache =
+          static_cast<tsi_ssl_session_cache*>(arg->value.pointer.p);
+    }
+  }
+  grpc_core::RefCountedPtr<grpc_channel_security_connector> sc =
+      grpc_tls_spiffe_channel_security_connector_create(
+          this->Ref(), std::move(call_creds), target_name,
+          overridden_target_name, ssl_session_cache);
+  if (sc == nullptr) {
+    return nullptr;
+  }
+  grpc_arg new_arg = grpc_channel_arg_string_create(
+      (char*)GRPC_ARG_HTTP2_SCHEME, (char*)"https");
+  *new_args = grpc_channel_args_copy_and_add(args, &new_arg, 1);
+  return sc;
+}
+
+grpc_tls_spiffe_server_credentials::grpc_tls_spiffe_server_credentials(
+    const grpc_tls_credentials_options* options)
+    : grpc_server_credentials(GRPC_CREDENTIALS_TYPE_SPIFFE),
+      options_(std::move(const_cast<grpc_tls_credentials_options*>(options))) {}
+
+grpc_tls_spiffe_server_credentials::~grpc_tls_spiffe_server_credentials() {}
+
+grpc_core::RefCountedPtr<grpc_server_security_connector>
+grpc_tls_spiffe_server_credentials::create_security_connector() {
+  return grpc_tls_spiffe_server_security_connector_create(this->Ref());
+}
+
+static bool CredentialOptionSanityCheck(
+    const grpc_tls_credentials_options* options) {
+  if (options == nullptr) {
+    gpr_log(GPR_ERROR, "SPIFFE TLS credentials options is nullptr.");
+    return false;
+  }
+  if (options->key_materials_config() == nullptr &&
+      options->credential_reload_config() == nullptr) {
+    gpr_log(
+        GPR_ERROR,
+        "SPIFFE TLS credentials options must specify either key materials or "
+        "credential reload config.");
+    return false;
+  }
+  return true;
+}
+
+grpc_channel_credentials* grpc_tls_spiffe_credentials_create(
+    const grpc_tls_credentials_options* options) {
+  if (!CredentialOptionSanityCheck(options)) {
+    gpr_log(GPR_ERROR,
+            "Invalid credentials options in creating TLS spiffe channel "
+            "credentials.");
+    return nullptr;
+  }
+  return grpc_core::New<grpc_tls_spiffe_credentials>(options);
+}
+
+grpc_server_credentials* grpc_tls_spiffe_server_credentials_create(
+    const grpc_tls_credentials_options* options) {
+  if (!CredentialOptionSanityCheck(options)) {
```

Comment:
For server credential, do you want to check if grpc_tls_credentials_options has server_authorization config, and return a warning/error?

Reply:
I added a warning, but did not fail the creation for that.